magda-authorization-api/src/NestedSetModelQueryer.ts

```diff
@@ -0,0 +1,1117 @@
+import * as pg from "pg";
+import * as _ from "lodash";
+const textTree = require("text-treeview");
+
+export interface NodeRecord {
+    [key: string]: any;
+}
+
+function isNonEmptyArray(v: any): boolean {
+    if (!v || !_.isArray(v) || !v.length) return false;
+    return true;
+}
+
+export type TextTreeNode =
+    | string
+    | {
+          text: string;
+          children?: TextTreeNode[];
+      };
+
+class NestedSetModelQueryer {
+    private pool: pg.Pool;
+    private tableName: string;
+
+    /**
+     * default select fields if [], all fields (i.e. `SELECT "id", "name"`) will be returned
+     *
+     * @type {string[]}
+     * @memberof NestedSetModelQueryer
+     */
+    public defaultSelectFieldList: string[] = ["id", "name"];
+
+    /**
+     * Default field list that will be used when insert nodes into tree.
+     * By default, only `name` field will be saved to database
+     * e.g. If your tree nodes have three properties (besides `id`, `left`, `right` --- they auto generated):
+     * - name
+     * - description
+     * - fullName
+     *
+     * Then you should set `defaultInsertFieldList` to ["name", "description", "fullName"]
+     *
+     * @type {string[]}
+     * @memberof NestedSetModelQueryer
+     */
+    public defaultInsertFieldList: string[] = ["name"];
+
+    /**
+     * Creates an instance of NestedSetModelQueryer.
+     * @param {pg.Pool} dbPool
+     * @param {string} tableName
+     * @param {string[]} [defaultSelectFieldList=null] default select fields; If null, all fields (i.e. `SELECT "id", "name"`) will be returned
+     * @memberof NestedSetModelQueryer
+     */
+    constructor(
+        dbPool: pg.Pool,
+        tableName: string,
+        defaultSelectFieldList: string[] = null,
+        defaultInsertFieldList: string[] = null
+    ) {
+        if (!dbPool) throw new Error("dbPool cannot be empty!");
+        if (!tableName) throw new Error("tableName cannot be empty!");
+
+        this.pool = dbPool;
+        this.tableName = tableName;
+
+        if (defaultSelectFieldList) {
+            if (!_.isArray(defaultSelectFieldList))
+                throw new Error("defaultSelectFieldList should be an array");
+
+            this.defaultSelectFieldList = defaultSelectFieldList;
+        }
+
+        if (defaultSelectFieldList) {
+            if (!_.isArray(defaultSelectFieldList))
+                throw new Error("defaultSelectFieldList should be an array");
+
+            this.defaultSelectFieldList = defaultSelectFieldList;
+        }
+
+        if (defaultInsertFieldList) {
+            if (!_.isArray(defaultInsertFieldList))
+                throw new Error("defaultInsertFieldList should be an array");
+
+            this.defaultInsertFieldList = defaultInsertFieldList;
+        }
+    }
+
+    private selectFields(
+        tableAliasOrName: string = "",
+        fields: string[] = null
```

Comment:
I think I was trying to make it generic so that the lib can be reused~ 😄
all table names & columns are actually hardcoded when creating the instance.  
I also made it now only allows `a-z, 0-9,  _` in any sql identifiers (i.e. table names, columns).
We shouldn't worry about SQL injection anymore.